devito/passes/iet/mpi.py

```diff
@@ -48,157 +47,129 @@ def _drop_halospots(iet):
-                not any(i & dep.cause for i in candidates) and
-                not any(i & loc_dims for i in candidates))
-
-    def rule1(dep, candidates, loc_dims):
-        # A reduction isn't a stopper to hoisting
-        return dep.write is not None and dep.write.is_reduction
-
-    rules = [rule0, rule1]
+    """
 
     # Precompute scopes to save time
     scopes = {i: Scope([e.expr for e in v]) for i, v in MapNodes().visit(iet).items()}
 
     # Analysis
     hsmapper = {}
     imapper = defaultdict(list)
-    for iters, halo_spots in MapNodes(Iteration, HaloSpot, 'groupby').visit(iet).items():
-        for hs in halo_spots:
-            hsmapper[hs] = hs.halo_scheme
-
-            for f, v in hs.fmapper.items():
-                loc_dims = frozenset().union([q for d in v.loc_indices
-                                              for q in d._defines])
-
-                for n, i in enumerate(iters):
-                    if i not in scopes:
-                        continue
-
-                    candidates = [i.dim._defines for i in iters[n:]]
-
-                    all_candidates = set().union(*candidates)
-                    reads = scopes[i].getreads(f)
-                    if any(set(a.ispace.dimensions) & all_candidates
-                           for a in reads):
-                        continue
-
-                    for dep in scopes[i].d_flow.project(f):
-                        if not any(r(dep, candidates, loc_dims) for r in rules):
-                            break
-                    else:
-                        hsmapper[hs] = hsmapper[hs].drop(f)
-                        imapper[i].append(hs.halo_scheme.project(f))
+
+    cond_mapper = _make_cond_mapper(iet)
+    iter_mapper = _filter_iter_mapper(iet)
+
+    for it, halo_spots in iter_mapper.items():
+        for hs0, hs1 in combinations(halo_spots, r=2):
+
+            if _check_control_flow(hs0, hs1, cond_mapper):
+                continue
+
+            # If there are overlapping loc_indices, skip
+            hs0_mdims = hs0.halo_scheme.loc_values
+            hs1_mdims = hs1.halo_scheme.loc_values
+            if hs0_mdims.intersection(hs1_mdims):
+                continue
+
+            for f, v in hs1.fmapper.items():
+
+                if f not in hs0.functions:
+                    continue
+
+                for dep in scopes[it].d_flow.project(f):
+                    if not any(r(dep, hs1, v.loc_indices) for r in rules):
                         break
+                else:
+                    # hs1 can be hoisted out of `it`, but we need to infer valid
+                    # loc_indices
+                    hse = hs1.halo_scheme.fmapper[f]
+                    raw_loc_indices = {}
+
+                    for d, v in hse.loc_indices.items():
+                        if v in it.uindices:
+                            v_sub = it.start
+                            raw_loc_indices[d] = v.symbolic_min.subs(it.dim, v_sub)
```

Comment:
what can v.symbolic_min be here?

Reply:
This is what happens here, in case `v` is not in uindices like the time_blocking example we use the already existing variable. In backwards cases `v_sub` is `time_M`.
```
(Pdb) print(v.symbolic_min)
Mod(time, 2)
(Pdb) print(v_sub)
time_m
(Pdb) print(raw_loc_indices)
{t: Mod(time_m, 2)}
```